altair/utils/_transformed_data.py

```diff
@@ -0,0 +1,545 @@
+    facet_mapping = get_facet_mapping(vega_spec)
+    dataset_mapping = get_datasets_for_chart_names(
+        vega_spec, chart_names, facet_mapping
+    )
+
+    # Build a list of vega dataset names that corresponds to the order
+    # of the chart components
+    dataset_names = []
+    for chart_name in chart_names:
+        if chart_name in dataset_mapping:
+            dataset_names.append(dataset_mapping[chart_name])
+        else:
+            raise ValueError("Failed to locate all datasets")
+
+    # Extract transformed datasets with VegaFusion
+    datasets, warnings = runtime.pre_transform_datasets(
+        vega_spec,
+        dataset_names,
+        get_local_tz(),
+        row_limit=row_limit,
+        inline_datasets=inline_datasets,
+    )
+
+    if isinstance(chart, (Chart, FacetChart)):
+        # Return DataFrame (or None if it was excluded) if input was a simple Chart
+        if not datasets:
+            return None
+        else:
+            return datasets[0]
+    else:
+        # Otherwise return the list of DataFrames
+        return datasets
+
+
+def make_magic_chart_name(i: int) -> str:
+    """Make magic chart name for chart number i
+
+    Parameters
+    ----------
+    i : int
+        Mark number
+
+    Returns
+    -------
+    str
+        Mark name
+    """
+    return MAGIC_CHART_NAME.format(i)
+
+
+def name_chart(
+    chart: Union[
+        Chart, FacetChart, LayerChart, HConcatChart, VConcatChart, ConcatChart
+    ],
+    i: int = 0,
+    exclude: Optional[Iterable[str]] = None,
+) -> List[str]:
+    """Name unnamed charts and subcharts
+
+    Name unnamed charts and subcharts so that we can look them up later in
+    the compiled Vega spec.
+
+    Note: This function mutates the input chart by applying names to
+    unnamed charts.
+
+    Parameters
+    ----------
+    chart : Chart, FacetChart, LayerChart, HConcatChart, VConcatChart, or ConcatChart
+        Altair chart to apply names to
+    i : int (default 0)
+        Starting chart index
+    exclude : iterable of str
+        Names of charts to exclude
+
+    Returns
+    -------
+    list of str
+        List of the names of the charts and subcharts
+    """
+    exclude = set(exclude) if exclude is not None else set()
+    if isinstance(chart, (Chart, FacetChart)):
+        # Perform shallow copy of chart so that we can change
+        # the name
+        if chart.name not in exclude:
+            if chart.name in (None, Undefined):
+                name = make_magic_chart_name(i)
+                chart.name = name
+            return [chart.name]
+        else:
+            return []
+    else:
+        if isinstance(chart, LayerChart):
+            subcharts = chart.layer
+        elif isinstance(chart, HConcatChart):
+            subcharts = chart.hconcat
+        elif isinstance(chart, VConcatChart):
+            subcharts = chart.vconcat
+        elif isinstance(chart, ConcatChart):
+            subcharts = chart.concat
+        else:
```

Comment:
Right now, `RepeatChart` doesn't have a `_transformed_data()` method, so it won't get this far. We could add a `_transformed_data()` method to `RepeatChart` that always raises an exception that it's not supported. But this might also be confusing (why have the method as all if it never works). What do you think?

Reply:
Just wondering, why can it never works? A `NotImplementedError` seems OK for me for now. It is also confusing if `RepeatChart` is the only `Chart` without a `transformed_data()` method.

Reply:
Yeah, a NotImplementedError sounds like a good approach. I'll add that.  I think it's probably possible to support `RepeatChart` in the future, I just haven't work through quite what it means, and how to process the structure of the Vega that Vega-Lite produces in this case.

Reply:
Done in a738408a8055bb9c7f3346d58411e4336d641346

Reply:
Now that all top-level charts have the `._transformed_data` method, you could add it to `TopLevelMixin` so there is no need anymore to repeat it.

Reply:
I'm not sure. The return type in the type signature is different between `Chart`/`FacetChart` (where it's `Optional[DataFrameLike]`, and the others (where it's `List[DataFrameLike]`.

For the standalone utility `transformed_data` function I used `typeing.overload` to express the two signatures. But I wasn't sure how to accomplish this in a superclass method when the type signature depends on the type of `self`.  Do you have any ideas on that?

Reply:
Good point. Just tried and got this far with defining _transformed_data on TopLevelMixin:

```python
   @overload
    def _transformed_data(
        self: Union["Chart", "FacetChart", "RepeatChart"],
        row_limit: Optional[int] = ...,
        exclude: Optional[Iterable[str]] = ...,
    ) -> Optional[DataFrameLike]:
        ...

    @overload
    def _transformed_data(
        self: Union["LayerChart", "ConcatChart", "HConcatChart", "VConcatChart"],
        row_limit: Optional[int] = ...,
        exclude: Optional[Iterable[str]] = ...,
    ) -> List[DataFrameLike]:
        ...

    def _transformed_data(
        self,
        row_limit=None,
        exclude=None,
    ):
```
But then mypy complains, rightly so, that these types are not superclasses of `TopLevelMixin`:
```
altair/vegalite/v5/api.py:2407: error: The erased type of self "Union[altair.vegalite.v5.api.Chart, altair.vegalite.v5.api.FacetChart, altair.vegalite.v5.api.RepeatChart]" is not a supertype of its class "altair.vegalite.v5.api.TopLevelMixin"  [misc]
altair/vegalite/v5/api.py:2415: error: The erased type of self "Union[altair.vegalite.v5.api.LayerChart, altair.vegalite.v5.api.ConcatChart, altair.vegalite.v5.api.HConcatChart, altair.vegalite.v5.api.VConcatChart]" is not a supertype of its class "altair.vegalite.v5.api.TopLevelMixin"  [misc]
```

According to [this GH issue](https://github.com/python/mypy/issues/7191#issuecomment-921570018) and some others, the solution might be to create protocols whcih identify the individual chart types but that seems to be more complicated then just keeping it as it is now -> I'm ok with leaving it.